### PR TITLE
[processing] New options to autofill batch processing dialog

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR reworks how Processing's batch dialog can be autopopulated. Previously, users could double-click a column header in order to "copy down" the top row's value to all other rows. This was very hidden, yet an important feature for users to utilise!

In this PR the dialog has been reworked to add a new "Autofill" row at the top of the table. Clicking on the Autofill option for any column opens a drop down menu with available fill options. The existing "fill down" option has been moved to this menu (so double-click header no longer works), and several new options have been added:

- "Calculate by expression":  Selecting it allows users to create a new QGIS expression to use to update all existing values within that column. Existing parameter values (including those from other columns) are available for use inside the expression via @variables. E.g. this allows setting output file names to complex expressions like ```'/home/me/stuff/buffer_' || left(@input, 30) || '_' || @distance || '.shp'```

- "Add values by expression": This option adds news rows using the values from an expression
which returns an array. (As opposed to "Calculate by Expression", which works only on existing rows). The intended use case is to allow populating the batch dialog using complex numeric series, e.g. those created by the `generate_series` expression function. For example, adding rows for a batch buffer using the expression `generate_series(100, 1000, 50)` results in new rows with values 100, 150, 200, .... 1000.

- "Add Files by Pattern…" (available for file/layer parameters only): Adds new rows to the table for matching files found using a file pattern and folder (with an option to search recursively)

Together, these options make the batch execution mode for Processing much more user friendly and powerful.

Sponsored by North Road, thanks to SLYR

### CODE REVIEW

File: src/gui/qgsfindfilesbypatternwidget.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * Returns the list of files found by the dialog.
+     * \see filesFound()
+     */
+    QStringList files() const { return mFiles; }
+
+  signals:
+
+    /**
+     * Emitted after files are found in the dialog.
```
Comment (nirvn) [inline src/gui/qgsfindfilesbypatternwidget.h]: Is it worth tweaking the documentation to make it clear that this signal will be emitted even if no files are found (with an empty string list)? 

I'm wondering whether, for clarity, the signal should be renamed to something like searchResult or something.

File: python/plugins/processing/gui/BatchPanel.py [hunk 2]
```diff
@@ -59,8 +91,182 @@
         os.path.join(pluginPath, 'ui', 'widgetBatchPanel.ui'))
 
 
-class BatchPanel(BASE, WIDGET):
+class BatchPanelFillWidget(QToolButton):
+
+    def __init__(self, parameterDefinition, column, panel, parent=None):
+        super().__init__(parent)
+
+        self.setBackgroundRole(QPalette.Window)
```
Comment (nirvn) [inline python/plugins/processing/gui/BatchPanel.py]: QPalette.Button might be a better color role to use here.

**Review by nirvn — COMMENTED** [review]
Code looking good.

File: src/gui/qgsfindfilesbypatternwidget.h (hunk 1 above)
Comment (nyalldawson) [reply src/gui/qgsfindfilesbypatternwidget.h]: Done, I've used findComplete and updated the dox

**Review by nirvn — APPROVED** [review]

**Pull request merged**